[rna-transcription approach]: a few improvements

petrem opened the pull request:
- Renamed `chr` to `char` in the snippets because `chr` is a built-in function and although shadowing it in this case may not be a problem, it is still a bad practice when it can be avoided.
- The dictionary-join approach mentions list comprehensions, but instead it uses a generator expression. Replaced this in the explanation and expanded to give the list comprehension based implementation along with a brief comparison.
- The overview mentions one approach is four times faster. In a brief comparison, it varies from 2.5x for a very short string and up to 60x faster for a 10^6 long one. Probably not worth going into the details, but 4x is just innacurate.

I noticed that the `maketrans` based approach mentions the translation is done between ASCII ordinal values. This happens to be true because the characters involved are ASCII letters, but the functions deal with unicode ordinals.

Do you think this is worth mentioning?

petrem committed:
rna-transcription approach: a few improvements

- Renamed `chr` to `char` in the snippets because `chr` is a built-in function and
  although shadowing it in this case may not be a problem, it is still a bad practice
  when it can be avoided.
- The dictionary-join approach mentions list comprehensions, but instead it uses a
  generator expression. Replaced this in the explanation and expanded to give the list
  comprehension based implementation along with a brief comparison.
- The overview mentions one approach is four times faster. In a brief comparison, it
  varies from 2.5x for a very short string and up to 60x faster for a 10^6 long one.
  Probably not worth going into the details, but 4x is just innacurate.

petrem committed:
several updates following discussions

- Replaced `char` with `nucleotide` as this is terminology from the domain.
- Rephrased a "see also" link to be more screen reader friendly.
- A note about the exercise not requiring tests for invalid characters is preset in one
  of the approaches. Copied it over to the other approach, for uniformity.
- Rephrased mention about performance and speedup.
- Replaced mention of ASCII with Unicode adding a brief explanation and links.

BethanyG commented:
Reopening to see if I can fix this.

BethanyG commented:
Hi @petrem - apologies for not replying earlier.  I am in PST, and went to bed on the early side last night.  🙂 

> _It's still a mess (that I've created) so I'd prefer to start afresh, if that's OK? Or perhaps there's something I'm missing and it's better to continue here, in that case, that's also fine._

If you are more comfortable restarting, let's do that.  I think the only issue is that this PR has quite a bit of discussion on it, so we lose that direct link.  But as long as you reference & link to this PR in your new one, I think we'll be fine.  😄

  > _I don't even know how I got here which makes it worse. Usually I simply use git cli but I don't find anything suspicious in the history. Must've done something really stupid in the editor <:-|_

Having just played with the gh cli last night (_it was the only way to edit the PR branch without creating a new one_), I **_think_** what happed was a [`gh pr update-branch`](https://cli.github.com/manual/gh_pr_update-branch) that was done directly from the repo (_not your fork_), and was missing the `--rebase` flag.  So it pulled in the last year's worth of changes on main, and merged them to your branch as "new" commits.  But I don't know for sure. 😱 

My typical workflow for anything I do on the track is:

**TL;DR?**  _Always pull in all upstream changes before branching to do work._ 😄 

1.   Switch to my local `main`, and do a `git fetch --all` to pull down all new refs.  I have remotes set for the repo (**upstream**) and my fork (**origin**).  
2. Once the refs download, I do a `git rebase upstream/main`, which replays my local `main` onto the fetched one.  (*e.g. it updates my main with all pulled in changes*)
3. After rebase, I check the status, and if needed, I push my local `main` changes to my fork (origin).
4. Once `main` has been updated, I checkout and update the branch via `git rebase main` - although typically, I am doing the update _before_ starting work, so instead of rebasing a branch I am creating one via `git checkout -b <branch name>`.
5. Once done with branch work, I do a `git push`, which updates my fork (origin).  From there, GH flags the change and the web interface asks if I want to PR to upstream.  If I've already PR'd, the upstream branch is updated automatically.

All that doesn't guarantee I won't  screw something up royally (_I have a talent_).  And if I do mess up, [oh shit, git](https://ohshitgit.com/) is usually my very first stop, followed by increasingly hysterical googling.  😆  

So - no harm, no foul!   I appreciate all the work and am happy to have you re-do the PR.  💙

petrem committed:
move note regarding testing for erroneous inputs to `introduction.md`
... because it applies to the exercise in general, not a particular approach.

Re-applying missed commits from prior cherry-pick.

BethanyG reviewed (approved):
Nice work, @petrem 🎉 

I am ready to merge this unless  @BNAndras & @kotp - either of you have additional comments or thoughts?

petrem committed:
convert code snippets to single quotes for consistency

petrem committed:
Re-applied the commits from below via cherry-pick.
convert code snippets to single quotes for consistency

petrem commented on exercises/practice/rna-transcription/.approaches/dictionary-join/content.md:
````diff
@@ -16,15 +16,32 @@ but the `LOOKUP` dictionary is defined with all uppercase letters, which is the
+A variant that uses a list comprehension is almost identical, but note the additional square brackets inside the `join()`:
+
+```python
+LOOKUP = {"G": "C", "C": "G", "T": "A", "A": "U"}
````
Converted to single quotes, thanks @IsaacG and @BethanyG . Unless someone else notices some more stuff that could be fixed, I'm done. Thank you.

>  But we've had enough of a pile-on with reviewers that I also won't complain if you want to merge this and move on.

As this is about improving the language and consistency of the approaches, I don't see why not - just didn't want to pile one more commit in case you started to do some work locally.

> I also hate to disappoint you, @petrem - my preferred quote style is single quotes, and IIRC the Python track docs somewhere say to go for single quotes over double. 😄 😆

Next you'll tell me your favorite color is not yellow, and your favorite editor is not Emacs?! Where is the world going... :wink:

BethanyG replied:
> _As this is about improving the language and consistency of the approaches, I don't see why not - just didn't want to pile one more commit in case you started to do some work locally._

Eh. Since I am out on (several) branches and am doing work intermittently, I am having to update as I go. Don't worry about bugging me ... unless of course you decided to merge 300+ commits in again.  I might tell you (_rather energetically_) to practice on a fork for that. 😉  😆   

>_Next you'll tell me your favorite color is not yellow, and your favorite editor is not Emacs?! Where is the world going... 😉_

The world is indeed in a sad state. 😉  But we are also here together doing good work with good cheer (_at least I hope so!_).  So something to celebrate.  🎉 ⭐ 😄

BethanyG commented:
> I can do a new branch and PR, of course, @BethanyG !

With a little more help from @IsaacG  and @kotp, we've got it fixed.  So I am going to merge and RUN  😄   But let's never, EVER do this again.  I learned a lot of great stuff, but am now loathe to practice it more.  Let's hope none of us ever get in this particular pickle again!  😆
